tests/approaches/test_grammar_search_invention_approach.py

```diff
@@ -61,7 +61,16 @@ def test_predicate_grammar():
          "NOT-((0:block).pose<=2.33), NOT-((0:block).width<=19.0)]")
     forall_grammar = _create_grammar("forall_single_feat_ineqs", dataset,
                                      env.predicates)
-    assert len(forall_grammar.generate(max_num=100)) == 100
+    # There are only so many unique predicates possible under the grammar.
+    # Non-unique predicates are pruned. Note that with a larger dataset,
+    # more predicates would appear unique.
+    assert len(forall_grammar.generate(max_num=10)) == 7
+    # Test CFG.grammar_search_predicate_cost_upper_bound.
+    default = CFG.grammar_search_predicate_cost_upper_bound
+    utils.update_config({"grammar_search_predicate_cost_upper_bound": 0})
+    assert len(single_ineq_grammar.generate(max_num=10)) == 0
+    # Reset to default just in case.
+    utils.update_config({"grammar_search_predicate_cost_upper_bound": default})
```

Comment:
Could you add a test that with an empty dataset, there should only ever be one predicate generated, since all predicates would have the same hash?